Source/NSString.m

```diff
@@ -1886,6 +1922,70 @@ - (void) getCharacters: (unichar*)buffer
     }
 }
 
+- (NSString *) stringByAddingPercentEncodingWithAllowedCharacters: (NSCharacterSet *)aSet
+{
+  NSData	*data = [self dataUsingEncoding: NSUTF8StringEncoding];
+  NSString	*s = nil;
+
+  if (data != nil)
+    {
+      unsigned char	*src = (unsigned char*)[data bytes];
+      unsigned int	slen = [data length];
+      unsigned char	*dst;
+      unsigned int	spos = 0;
+      unsigned int	dpos = 0;
+
+      dst = (unsigned char*)NSZoneMalloc(NSDefaultMallocZone(), slen * 3);
+      while (spos < slen)
+	{
+	  unichar	c = src[spos++];
+	  unsigned int	hi;
+	  unsigned int	lo;
+
+	  if([aSet characterIsMember: c]) // if the character is in the allowed set, put it in
+	    {
+	      dst[dpos++] = c;
+	    }
+	  else // if not, then encode it...
+	    {
+	      dst[dpos++] = '%';
+	      hi = (c & 0xf0) >> 4;
+	      dst[dpos++] = (hi > 9) ? 'A' + hi - 10 : '0' + hi;
+	      lo = (c & 0x0f);
+	      dst[dpos++] = (lo > 9) ? 'A' + lo - 10 : '0' + lo;
+	    }
+	}
+      s = [[NSString alloc] initWithBytes: dst
+				   length: dpos
+				 encoding: NSUTF8StringEncoding];
+      NSZoneFree(NSDefaultMallocZone(), dst);
+      IF_NO_GC([s autorelease];)
+    }
+  return s;
+}
+
+- (NSString *) stringByRemovingPercentEncoding
+{
+  NSData *data = [self dataUsingEncoding: NSUTF8StringEncoding];
+  NSString *result = nil;
+  char *s = (char *)[data bytes];
+  NSUInteger slen = 0; 
+  char *o = NULL;
+
+  // Allocate memory...
+  slen = strlen(s);
+  o = (char *)NSZoneMalloc(NSDefaultMallocZone(), slen * 3);
```

Comment:
Why three times the size of the original? The decoded length should always be at most the encoded one.

Reply:
Fixed